File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int right = 0;
        int maxLength = -1;
        Map<Character,Integer> lastIndex = new HashMap<>();
        for(right = 0; right < s.length(); right++){
            char rightChar = s.charAt(right);
            if(lastIndex.containsKey(rightChar)){
                left = Math.max(left,lastIndex.get(rightChar)+1);
            }
            lastIndex.put(rightChar,right);
            maxLength = Math.max(maxLength,right-left +1);
        }
        return maxLength;
        
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int maxLen =0;
        set<char> set;
        int left =0;
        for(int right =0 ;right<n; right++){
            while(set.contains(s[right]))
            {
                set.erase(s[left]);
                left++;
            }
            set.insert(s[right]);
            maxLen = max(maxLen,right-left+1);
        }
        return maxLen;
        
    }
};